Improved `GridLayout` docs. (#6368)
slint-ui/slint: docs/reference/src/language/builtins/elements.md

## `GridLayout`

`GridLayout` places elements on a grid.

It adds the following properties to each cell element: `col`, `row`, `colspan`, `rowspan`.
These properties must be compile-time constants.

`col` and `row` control the position of a cell on the grid structure.
`row` must be specified explicitly to be able to advance it&mdash;or it'll stay the same&mdash;which then implicitly resets an unspecified `col` to `0`.
Wrapping elements in `Row` elements instead will take care of row advancement for you.

> Omitting `col` implicitly means the next column (or the first at the beginning).

Author: > The phrase inside the dash is typically an emphasis, but here it describes an opposite.

The em dashes are what we in German call "Gedankenstrich". It has nothing to do with emphasizing.

> ```
>         Button { row: 1; col: 0; text: "Third"; } // new row, implicitly resets column to 0
> ```

It's not implicit, if it's explicit. 😉

I'm incorporating your suggestions into my own new commit, and further improving it.

> If the cell elements' size constraints don't work against this, a `GridLayout` covers its whole surface with its cells. The elements constituting the cells will be stretched inside their allotted space.

Author: It's something you have to learn about `GridLayout` (and something that maybe should change by supporting alignment on both axes), because other layouts like `VerticalLayout` allow you to align the whole block of elements at one side. The very last sentence with "allotted space" is the part that made me edit this section of the docs to add it.

In my app, e.g., I have form widgets in a `GridLayout`, so the text boxes on the right are nicely aligned with each other, and need to wrap it in a `VerticalLayout` with `alignment: start;` to be able to present the form properly and not stretched.

> If the cell elements' size constraints don't work against this, a `GridLayout` covers its whole surface with its cells. The elements constituting the cells will be stretched inside their allotted space.

Reviewer: Thanks for the explanation. Mentioning alignment etc. makes sense.

### Properties

-   **`spacing`** (_in_ _length_): The distance between the elements in the layout.
-   **`spacing-horizontal`**, **`spacing-vertical`** (_in_ _length_):
    Set these properties to override the spacing on specific axes.

> -   **`padding`** (_in_ _length_): The padding around the grid structure as a whole.

Reviewer: Nice improvement!

## `GridLayout`

`GridLayout` places elements on a grid.

> It adds the following properties to each cell element: `col`, `row`, `colspan`, `rowspan`.

Reviewer: On second thought, the "it adds" is a bit weird. How about this?

```suggestion
Cell elements inside a `GridLayout` obtain the following new properties:

 -   * **row** (_in_ _int_): The element's row within the grid. Set this property to override the sequential row assignment. Setting this property resets the element's column to zero, unless explicitly set.
-   * **col** (_in_ _int_): The element's column within the grid. Set this property to override the sequential column assignment.
-   * **colspan** (_in_ _int_): The number of columns this element should span. (default value: `1`)
-   * **rowspan** (_in_ _int_): The number of row this element should span. (default value: `1`)
 ```

This then also replaces the more verbose text further down.

Author: > Set this property to override the sequential row assignment.

You only have sequential row assignment with `Row`, right? Testing this in SlintPad, you can override the `Row`-assigned row indices with `row`, but this seems like an outlandish use case and should perhaps become an error.

> These properties must be compile-time constants.

Reviewer: ```suggestion
Any bindings to these properties must be compile-time constants.
```

`col` and `row` control the position of a cell within the grid.
Without `col`, elements are assigned columns sequentially, within their row.
Set the `row` property to place an element into a specific row. This implicitly resets the column to `0`.
Wrap elements in `Row` to avoid having to repeatedly specify row indices.

The following example creates a 2-by-2 grid with `Row` elements, omitting one cell:

```slint
GridLayout {
    Row { // children implicitly on row 0
        Button { col: 1; text: "Top Right"; } // implicit column after this would be 2
    }
    Row { // children implicitly on row 1
        Button { text: "Bottom Left"; }  // implicitly in column 0...
        Button { text: "Bottom Right"; } // ...and 1
    }
}
```

The following example creates the same grid with manual placement:

```slint
GridLayout {
    Button { row: 0; col: 1; text: "Top Right"; } // `row: 0;` could even be left out at the start
    Button { row: 1; text: "Bottom Left"; } // new row, implicitly resets column to 0
    Button { text: "Bottom Right"; } // same row, sequentially assigned column 1
}
```

> Since a `GridLayout` doesn't align, it covers its entire surface with its cells. The elements constituting the cells will be stretched inside their allotted space. (The only exception being that the cell elements' size constraints like, e.g., `min-width` or `max-width` could work against this.)

Reviewer: ```suggestion
`GridLayout` covers its entire surface with cells. Cells are not aligned. The elements constituting the cells will be stretched inside their allocated space, unless their size constraints, like, e.g. `min-width` or `max-width`, could work against this.
```
